### ISSUE TITLE
Unexpected behaviour for memory layout of discriminants in enums

### ISSUE THREAD

**Issue by graymalkin** [rust-lang/rust#93773]
Hi, I've been looking at the implementation of enums and their memory layout. I've noticed something which surprised me and seems a little inconsistent. It's hard to tell if this behaviour is good/neutral/bug, so apologies if this is not the right place to raise it.

```rs
#![feature(rustc_attrs)]

#[rustc_layout(debug)]
enum E {
    A,
    B,
    C,
}
```

Looking at the layout information for this struct, I can see that the discriminant is laid out as an `I8` type:

```rs
layout_of(E) = Layout {
           fields: Arbitrary {
               offsets: [
                   Size {
                       raw: 0,
                   },
               ],
               memory_index: [
                   0,
               ],
           },
           variants: Multiple {
               tag: Scalar {   // <-- The discriminant
                   value: Int(
                       I8,     // I8 storage
                       false,
                   ),
                   valid_range: 0..=2,
               },
               tag_encoding: Direct,
               tag_field: 0,
               variants: [ ... ],
           },
           abi: Scalar(
               Scalar {
                   value: Int(
                       I8,
                       false,
                   ),
                   valid_range: 0..=2,
               },
           ),
           ...
       }
```

This seems sensible, the discriminant can have one of 3 values for `E::A` `E::B` or `E::C`.

If I modify the enum so that one of the members stores a `u64` then the layout of the discriminant appears to change:

```rs
#![feature(rustc_attrs)]

#[rustc_layout(debug)]
enum E {
    A(u64),
    B,
    C,
}
```

Now the discriminant is laid out as an I64. Why?

```rs
layout_of(E) = Layout {
           fields: Arbitrary {
               offsets: [
                   Size {
                       raw: 0,
                   },
               ],
               memory_index: [
                   0,
               ],
           },
           variants: Multiple {
               tag: Scalar {  // <-- Discriminant
                   value: Int(
                       I64,   // I64 storage?
                       false,
                   ),
                   valid_range: 0..=2,
               },
               tag_encoding: Direct,
               tag_field: 0,
               variants: [ ... ],
           },
           abi: ScalarPair(
               Scalar {
                   value: Int(
                       I64,
                       false,
                   ),
                   valid_range: 0..=2,
               },
               Scalar {
                   value: Int(
                       I64,
                       false,
                   ),
                   valid_range: 0..=18446744073709551615,
               },
           ),
           ...
       }
```

One final modification, if I make `E::B` store a `u32` the discriminant storage changes again:

```rs
#![feature(rustc_attrs)]

#[rustc_layout(debug)]
enum E {
    A(u64),
    B(u32),
    C,
}
```

Now the discriminant is stored in an `I32`:

```rs
layout_of(E) = Layout {
           fields: Arbitrary {
               offsets: [
                   Size {
                       raw: 0,
                   },
               ],
               memory_index: [
                   0,
               ],
           },
           variants: Multiple {
               tag: Scalar {  // <-- Discriminant
                   value: Int(
                       I32,   // I32 storage?
                       false,
                   ),
                   valid_range: 0..=2,
               },
               tag_encoding: Direct,
               tag_field: 0,
               variants: [ ... ],
           },
           abi: Aggregate {
               sized: true,
           },
           largest_niche: Some(
               Niche {
                   offset: Size {
                       raw: 0,
                   },
                   scalar: Scalar {
                       value: Int(
                           I32,
                           false,
                       ),
                       valid_range: 0..=2,
                   },
               },
           ),
           ...
```

It appears that the discriminant is stored in storage that is the minimum of the sizes of the non-empty members of the enum. Why is it not always stored in a storage type which is large enough to cover the range of the discriminant, or preferentially something of the platform's word size?

Thanks,
Simon

**Comment by nagisa** [comment]
The `u32` has an alignment of 4 and `u64` has an alignment of 8 bytes (typically). So for the type to satisfy these alignment requirements there would either need to be padding, or the discriminant itself can simply become larger, which in this case it does.

**Comment by RalfJung** [comment]
> So for the layout to satisfy these alignment requirements there would either need to be padding, or the discriminant itself can simply become larger, 

Agreed -- but padding seems like the more obvious choice, so this doesn't exactly explain how the discriminant type comes to be.

**Comment by jyn514** [comment]
Is this a bug? It's unclear to me what the goal of the issue is - is there some more efficient layout we could use?

**Comment by RalfJung** [comment]
I don't think it's a bug, and we don't guarantee any particular layout for this enum. So while I also can't explain why we choose I64 over I8, I also don't think we need to track this as an issue.

> It appears that the discriminant is stored in storage that is the minimum of the sizes of the non-empty members of the enum. Why is it not always stored in a storage type which is large enough to cover the range of the discriminant, or preferentially something of the platform's word size?

Maybe @oli-obk or @eddyb knows. But it might also just be something that "just happened" and since there's no real upside or downside either way there is no reason to change it.

**Comment by eddyb** [comment]
> I don't think it's a bug, and we don't guarantee any particular layout for this enum. So while I also can't explain why we choose I64 over I8, I also don't think we need to track this as an issue.

I agree on the first part, but see below for why the second part is trickier.

> Why is it not always stored in a storage type which is large enough to cover the range of the discriminant, **_or preferentially something of the platform's word size?_**

<sup>*(emphasis mine)*</sup>

The maximum alignment of an integer is already typically the same as the native general-purpose register size.
That's what you're seeing when it picks `I64` <sub>(i.e. it shouldn't be able to pick `I128` on that target - if you can trick it to do that, e.g. by using SIMD fields, maybe open an issue about that specific aspect, because `I128` and `I64` should almost always have the same alignment, i.e. 4 or 8 depending on the target, and when it's `8` we should pick `I64`, not `I128`, even if the latter *also* fits in the padding gap)</sub>.

And if were to always pick the maximum, that could increase the size of the `enum`, which we wouldn't want to.
Consider this `enum`, which also has the discriminant "grow into the padding" (to become `I32`):
```rust
enum E {
    A(u64),
    B(u32, u32, u32),
    C,
}
```

> Agreed -- but padding seems like the more obvious choice, so this doesn't exactly explain how the discriminant type comes to be.

Eh, padding is *not* that clear of a choice, as it requires more advanced LLVM optimizations to reduce the costs of going back and forth between registers and in-memory representation, esp. when combined with `memcpy` etc.

"Growing into the padding" is an easy way to reduce the complexity of the layout in many cases, and that tends to lead to shorter compile times, better codegen quality, etc.

*However*, this decision predates not only the modern layout system, but 1.0 itself - it may have been in `rustc` already by 2014 (but I can't remember for sure), and a lot has changed about LLVM in that time.

It's possible we could just remove that logic and not pay for it *but there's a catch*.
Not only are there codegen implications, but the tag is the only niche a "tagged union" `enum` can have.

By reclaiming space wasted by the padding, we can ensure you always have a larger niche range than you can easily use up. <sub>(though I'll admit even `u8` is hard to fully drain, you would need a lot of `Either` nesting, which has only in the last year or so has it become possible to do without increasing the size at every step, thanks to the "packing fields around niche" `enum` layout that's finally been implemented)</sub>

**Comment by graymalkin** [comment]
Can this rationale be documented in the implementation somewhere? This explanation is elucidating and it would be a pity to loose it.

**Comment by jyn514** [comment]
@graymalkin are you interested in making a PR adding it as a comment? :)

**Comment by graymalkin** [comment]
Sure, I’ll try and find an appropriate place around the layout code.
